core/trino-main/src/main/java/io/trino/execution/scheduler/HashDistributionSplitAssigner.java

```diff
@@ -93,33 +120,43 @@ public AssignmentResult assign(PlanNodeId planNodeId, ListMultimap<Integer, Spli
             }
         }
         else {
-            for (Integer outputPartitionId : splits.keySet()) {
+            splits.forEach((outputPartitionId, split) -> {
                 TaskPartition taskPartition = outputPartitionToTaskPartition.get(outputPartitionId);
                 verify(taskPartition != null, "taskPartition not found for outputPartitionId: %s", outputPartitionId);
-                if (!taskPartition.isIdAssigned()) {
-                    // Assigns lazily to ensure task ids are incremental and with no gaps.
-                    // Gaps can occur when scanning over a bucketed table as some buckets may contain no data.
-                    taskPartition.assignId(nextTaskPartitionId++);
+
+                List<SubPartition> subPartitions;
+                if (taskPartition.getSplitBy().isPresent() && taskPartition.getSplitBy().get().equals(planNodeId)) {
+                    subPartitions = ImmutableList.of(taskPartition.getNextSubPartition());
                 }
-                int taskPartitionId = taskPartition.getId();
-                if (!createdTaskPartitions.contains(taskPartitionId)) {
-                    Set<HostAddress> hostRequirement = sourcePartitioningScheme.getNodeRequirement(outputPartitionId)
-                            .map(InternalNode::getHostAndPort)
-                            .map(ImmutableSet::of)
-                            .orElse(ImmutableSet.of());
-                    assignment.addPartition(new Partition(
-                            taskPartitionId,
-                            new NodeRequirements(catalogRequirement, hostRequirement)));
-                    for (PlanNodeId replicatedSource : replicatedSplits.keySet()) {
-                        assignment.updatePartition(new PartitionUpdate(taskPartitionId, replicatedSource, replicatedSplits.get(replicatedSource), completedSources.contains(replicatedSource)));
-                    }
-                    for (PlanNodeId completedSource : completedSources) {
-                        assignment.updatePartition(new PartitionUpdate(taskPartitionId, completedSource, ImmutableList.of(), true));
+                else {
+                    subPartitions = taskPartition.getSubPartitions();
+                }
+
+                for (SubPartition subPartition : subPartitions) {
+                    if (!subPartition.isIdAssigned()) {
+                        int taskPartitionId = nextTaskPartitionId++;
+                        // Assigns lazily to ensure task ids are incremental and with no gaps.
```

Comment:
Also:
should you check if id is not already assigned in 
```
public void assignId(int id)
```

Reply:
> remind me. Why gaps are bad?

In fault tolerant tests we expect `partition 0` to always be present, as that's where we injecting a failure. Tests become flaky when partition 0 is missing for whatever reason.

I considered fixing the tests themselves, but it wasn't trivial. So than I thought that it might be easier to simply ensure there's no gaps. Also it will be less confusing from the end user perspective, as it may not be clear if missing tasks is a bug of it works as expected.